langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiGeminiStreamingChatModel.java

```diff
@@ -0,0 +1,228 @@
+                .tools(FunctionMapper.fromToolSepcsToGTool(toolSpecifications, this.allowCodeExecution))
+                .toolConfig(new GeminiToolConfig(this.toolConfig))
+                .build();
+    }
+
+    private ChatModelRequest createChatModelRequest(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
+        return ChatModelRequest.builder()
+                .model(modelName)
+                .temperature(temperature)
+                .topP(topP)
+                .maxTokens(maxOutputTokens)
+                .messages(messages)
+                .toolSpecifications(toolSpecifications)
+                .build();
+    }
+
+    private void processGenerateContentRequest(GeminiGenerateContentRequest request, StreamingResponseHandler<AiMessage> handler,
+                                               ChatModelRequest chatModelRequest, ConcurrentHashMap<Object, Object> listenerAttributes) {
+        GeminiStreamingResponseBuilder responseBuilder = new GeminiStreamingResponseBuilder(this.includeCodeExecutionOutput);
+
+        try {
+            Stream<GeminiGenerateContentResponse> contentStream = withRetry(
+                    () -> this.geminiService.generateContentStream(this.modelName, this.apiKey, request),
+                    maxRetries);
+
+            contentStream.forEach(partialResponse -> {
+                Optional<String> text = responseBuilder.append(partialResponse);
+                text.ifPresent(handler::onNext);
+            });
+
+            Response<AiMessage> fullResponse = responseBuilder.build();
+            handler.onComplete(fullResponse);
+
+            notifyListenersOnResponse(fullResponse, chatModelRequest, listenerAttributes);
+        } catch (RuntimeException exception) {
+            notifyListenersOnError(exception, chatModelRequest, listenerAttributes);
+            handler.onError(exception);
+        }
+    }
+
+    private static String computeMimeType(ResponseFormat responseFormat) {
+        if (responseFormat == null || ResponseFormatType.TEXT.equals(responseFormat.type())) {
+            return "text/plain";
+        }
+
+        if (ResponseFormatType.JSON.equals(responseFormat.type()) &&
+                responseFormat.jsonSchema() != null &&
+                responseFormat.jsonSchema().rootElement() != null &&
+                responseFormat.jsonSchema().rootElement() instanceof JsonEnumSchema) {
+            return "text/x.enum";
+        }
+
+        return "application/json";
+    }
+
+    private void notifyListenersOnRequest(ChatModelRequestContext context) {
+        listeners.forEach((listener) -> {
+            try {
+                listener.onRequest(context);
+            } catch (Exception e) {
+                log.warn("Exception while calling model listener (onRequest)", e);
+            }
+        });
+    }
+
+    private void notifyListenersOnResponse(Response<AiMessage> fullResponse, ChatModelRequest chatModelRequest,
```

Comment:
This would require a more extensive refactor of the GoogleAiGeminiChatModel class, but I can certainly do it. However, the same question as above arises: which class should contain the logic? A new class? Perhaps even an abstract one?